src/NextcloudCalendar.php

```diff
@@ -0,0 +1,443 @@
+     */
+    final public function getVisibility($path)
+    {
+        return $this->getMetadata($path);
+    }
+
+    /**
+     * Check whether a file exists.
+     *
+     * @param string $path
+     *
+     * @return bool
+     */
+    final public function has($path)
+    {
+        if ($path == ".acl" && $this->defaultAcl) {
+            return true;
+        }
+
+        $calendarId = $this->getCalendarId($path);
+        if ($calendarId) {
+            return true;
+        } else {
+            $filename = basename($path);
+            $calendar = dirname($path);
+            $calendarId = $this->getCalendarId($calendar);
+            $calendarItem = $this->calDavBackend->getCalendarObject($calendarId, $filename);
+            if ($calendarItem) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    /**
+     * List contents of a directory.
+     *
+     * @param string $directory
+     * @param bool $recursive
+     *
+     * @return array
+     */
+    final public function listContents($directory = '', $recursive = false)
+    {
+        $result = [];
+        if ($directory === "") {
+            $calendars = $this->calDavBackend->getCalendarsForUser($this->userId);
+            $result = array_map(function ($calendar) {
+                return $this->normalizeCalendar($calendar);
+            }, $calendars);
+
+            return $result;
+        } else {
+            $directory = basename($directory);
+
+            $calendar = $this->calDavBackend->getCalendarByUri($this->principalUri, $directory);
+    	    $calendarObjects = $this->calDavBackend->getCalendarObjects($calendar['id']);
+    	    $contents = [];
+
+    	    foreach ($calendarObjects as $calendarObject) {
+                $contents[] = $this->calDavBackend->getCalendarObject($calendarObject['calendarid'], $calendarObject['uri']);
+            }
+	        $result = array_map(function($calendarItem) use ($directory) {
+                return $this->normalizeCalendarItem($calendarItem, $directory);
+	        }, $contents);
+    	    return $result;
+        }
+    }
+
+    /**
+     * Read a file.
+     *
+     * @param string $path
+     *
+     * @return array|false
+     */
+    final public function read($path)
+    {
+        if ($path == ".acl" && $this->defaultAcl) {
+            return $this->normalizeAcl($this->defaultAcl);
+        }
+
+        $filename = basename($path);
+        $calendar = dirname($path);
+        $calendarId = $this->getCalendarId($calendar);
+        $calendarItem = $this->calDavBackend->getCalendarObject($calendarId, $filename);
+        if (!$calendarItem) {
+            return false;
+        }
+        return $this->normalizeCalendarItem($calendarItem, $calendar);
+    }
+
+    /**
+     * Read a file as a stream.
+     *
+     * @param string $path
+     *
+     * @return array|false
+     */
+    final public function readStream($path)
+    {
+        return false;
```

Comment:
In other implementations where no stream is supported, we've just called `return $this->read($path);` here...
Is that a good idea?  :shrug:

Reply:
agreed!